src/codegen/transformations.lisp

```diff
@@ -7,208 +7,200 @@
    #:coalton-impl/algorithm
    #:immutable-map-data)
   (:local-nicknames
+   (#:parser #:coalton-impl/parser)
    (#:util #:coalton-impl/util)
    (#:tc #:coalton-impl/typechecker))
   (:export
+   #:action
    #:traverse
-   #:traverse-bindings
+   #:traverse-with-binding-list
    #:update-function-env
-   #:make-function-t))
+   #:make-function-t
+   #:rename-type-variables
+   #:node-variables
+   #:node-free-p))
 
 (in-package #:coalton-impl/codegen/transformations)
 
-(defun traverse-bindings (bindings funs)
-  (declare (type binding-list bindings)
-           (values binding-list))
-  (loop :for (name . node) :in bindings
-        :collect (cons name (traverse node funs nil))))
+(defun node-subtype-p (symbol)
+  (and (symbolp symbol) (subtypep symbol 'node)))
+
+(deftype node-subtype ()
+  '(satisfies node-subtype-p))
+
+(deftype when-keyword ()
+  '(member :before :after :traverse))
+
+(defstruct (action (:constructor make-action (when type function)))
+  "Container for a function to call on nodes of a particular type - before, after, or instead of the recursive call in the traversal"
+  (when     ':after                   :type when-keyword :read-only t)
+  (type     'node                     :type node-subtype :read-only t)
+  (function (util:required 'function) :type function     :read-only t))
+
+(defmacro action ((when type &rest args) &body body)
+  "Construct an action block, turning `body` into a lambda function accepting `args` if provided"
```

Comment:
Since macros are new syntax, you must write something comprehensive about the syntax in the doc string.